### PR TITLE
[Feature] lake pk table compaction

### PR DESCRIPTION
**Pull request by luohaha**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #14853

## Problem Summary(Required) ：

Include the implementation of compaction logic in lake primary key table.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by luohaha** [comment]
run starrocks_be_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/storage/lake/compaction_policy.cpp [hunk 1]
```diff
@@ -37,6 +39,93 @@ class BaseAndCumulativeCompactionPolicy : public CompactionPolicy {
...
+    using RowsetCandidate = std::pair<RowsetMetadataPtr, RowsetStat>;
+    explicit PrimaryCompactionPolicy(TabletPtr tablet) : _tablet(std::move(tablet)) {}
+    ~PrimaryCompactionPolicy() override = default;
+
+    StatusOr<std::vector<RowsetPtr>> pick_rowsets(int64_t version) override;
+
+private:
+    double calc_compaction_score(const RowsetStat& stats);
+    static const size_t compaction_result_bytes_threashold = 1000000000;
+    static const size_t compaction_result_rows_threashold = 10000000;
```
Comment (sduzh) [inline be/src/storage/lake/compaction_policy.cpp:59]:
```suggestion
    static const size_t kCompactionResultBytesThreashold = 1000000000;
    static const size_t kCompactionResultRowsThreashold = 10000000;
    
    double calc_compaction_score(const RowsetStat& stats);

```

File: be/src/storage/lake/compaction_policy.cpp [hunk 2]
```diff
@@ -37,6 +39,93 @@ class BaseAndCumulativeCompactionPolicy : public CompactionPolicy {
...
+            if (delete_bytes_a < delete_bytes_b) {
+                return true;
+            } else if (delete_bytes_a > delete_bytes_b) {
+                return false;
+            } else {
+                // may happen when deleted rows is zero
+                return a.second.bytes > b.second.bytes;
+            }
+        }
+    };
```
Comment (sduzh) [inline be/src/storage/lake/compaction_policy.cpp:91]: Nit: how about define `RowsetCandidate` as a new class and make a custom '<' operator of it?

File: be/src/storage/lake/compaction_policy.cpp [hunk 3]
```diff
@@ -59,6 +148,16 @@ double base_compaction_score(const TabletMetadataPB& metadata) {
     return cumulative_point - 1;
 }
 
+double primary_compaction_score(const TabletMetadataPB& metadata) {
+    uint32_t segment_num_score = 0;
+    for (uint32_t i = 0; i < metadata.rowsets_size(); i++) {
+        const auto& rowset = metadata.rowsets(i);
+        segment_num_score += rowset.overlapped() ? rowset.segments_size() : 1;
+    }
+    LOG(INFO) << fmt::format("tablet: {}, primary compaction score: {}", metadata.id(), segment_num_score);
```
Comment (sduzh) [inline be/src/storage/lake/compaction_policy.cpp:157]: Remove this log?

File: be/src/storage/lake/delta_writer.cpp [hunk 4]
```diff
@@ -285,13 +279,18 @@ Status DeltaWriterImpl::finish() {
     op_write->mutable_rowset()->set_data_size(_tablet_writer->data_size());
     op_write->mutable_rowset()->set_overlapped(op_write->rowset().segments_size() > 1);
     // handle partial update
-    if (_rowset_txn_meta != nullptr && _partial_update_tablet_schema != nullptr) {
-        op_write->mutable_txn_meta()->CopyFrom(*_rowset_txn_meta);
+    RowsetTxnMetaPB* rowset_txn_meta = _tablet_writer->rowset_txn_meta();
+    if (rowset_txn_meta != nullptr && _partial_update_tablet_schema != nullptr) {
```
Comment (sduzh) [inline be/src/storage/lake/delta_writer.cpp:283]: Question: is it possible to have `rowset_txn_meta != nullptr && _partial_update_tablet_schema == nullptr` or `rowset_txn_meta == nullptr && _partial_update_tablet_schema != nullptr`

File: be/src/storage/lake/compaction_policy.cpp [hunk 5]
```diff
@@ -37,6 +39,93 @@ class BaseAndCumulativeCompactionPolicy : public CompactionPolicy {
...
+    std::stringstream input_infos;
+    while (!rowset_queue.empty()) {
+        const auto& rowset_candidate = rowset_queue.top();
+        cur_compaction_result_bytes += rowset_candidate.second.bytes;
+        cur_compaction_result_rows += rowset_candidate.second.num_rows;
+        if (input_rowsets.size() > 0 && ((cur_compaction_result_bytes > compaction_result_bytes_threashold * 3 / 2) ||
+                                         (cur_compaction_result_rows > compaction_result_rows_threashold * 3 / 2))) {
+            break;
+        }
+        input_rowsets.emplace_back(std::make_shared<Rowset>(_tablet.get(), std::move(rowset_candidate.first), i++));
```
Comment (sduzh) [inline be/src/storage/lake/compaction_policy.cpp:113]: Pass `i` as the rowset index may cause unexpected behaviors.

File: be/src/storage/lake/delta_writer.cpp [hunk 6]
```diff
@@ -285,13 +279,18 @@ Status DeltaWriterImpl::finish() {
...
+        op_write->mutable_txn_meta()->CopyFrom(*rowset_txn_meta);
         for (auto i = 0; i < _partial_update_tablet_schema->columns().size(); ++i) {
             const auto& tablet_column = _partial_update_tablet_schema->column(i);
             op_write->mutable_txn_meta()->add_partial_update_column_ids(_referenced_column_ids[i]);
             op_write->mutable_txn_meta()->add_partial_update_column_unique_ids(tablet_column.unique_id());
         }
+        // generate rewrite segment names to avoid gc in rewrite operation
+        for (auto i = 0; i < op_write->rowset().segments_size(); i++) {
+            op_write->add_rewrite_segments(fmt::format("{}.dat", generate_uuid_string()));
```
Comment (sduzh) [inline be/src/storage/lake/delta_writer.cpp:292]:
```suggestion
            op_write->add_rewrite_segments(random_segment_filename());
```

File: be/src/storage/lake/delta_writer.cpp (hunk 4 above)
Comment (luohaha) [reply be/src/storage/lake/delta_writer.cpp:283]: `rowset_txn_meta` can't be nullptr, if it is primary table. I will use CHECK to make sure this. And _partial_update_tablet_schema can be nullptr when it is normal update, or not nullptr when it is partial update.

File: be/src/storage/lake/delta_writer.cpp (hunk 6 above)
Comment (luohaha) [reply be/src/storage/lake/delta_writer.cpp]: Done

File: be/src/storage/lake/compaction_policy.cpp (hunk 1 above)
Comment (luohaha) [reply be/src/storage/lake/compaction_policy.cpp]: Done

File: be/src/storage/lake/compaction_policy.cpp (hunk 5 above)
Comment (luohaha) [reply be/src/storage/lake/compaction_policy.cpp]: Done

File: be/src/storage/lake/compaction_policy.cpp (hunk 2 above)
Comment (luohaha) [reply be/src/storage/lake/compaction_policy.cpp]: Done

File: be/src/storage/lake/tablet_manager.cpp [hunk 7]
```diff
@@ -610,18 +621,26 @@ Status apply_txn_log(const TxnLog& log, TabletMetadata* metadata) {
...
+        RETURN_IF_ERROR(apply_pk_compaction_log(log.op_compaction(), metadata, tablet, builder, base_version));
     }
     return Status::OK();
 }
 
 StatusOr<double> publish(LocationProvider* location_provider, Tablet* tablet, int64_t base_version, int64_t new_version,
                          const int64_t* txns, int txns_size) {
     auto compaction_score = [](const TabletMetadata& metadata) {
-        return std::max(base_compaction_score(metadata), cumulative_compaction_score(metadata));
+        if (is_primary_key(metadata)) {
```
Comment (sduzh) [inline be/src/storage/lake/tablet_manager.cpp:639]:
```suggestion
        return compaction_score(metadata);
```

File: be/src/storage/lake/lake_primary_index.cpp [hunk 8]
```diff
@@ -100,6 +106,9 @@ Status LakePrimaryIndex::_do_lake_load(Tablet* tablet, TabletMetadata* metadata,
         }
     }
     _tablet_id = tablet->id();
+    if (watch.elapsed_time() > 10 * 1000 * 1000) {
```
Comment (sduzh) [inline be/src/storage/lake/lake_primary_index.cpp:109]:
```suggestion
    if (watch.elapsed_time() > /*10ms=*/10 * 1000 * 1000) {
```

File: be/src/storage/lake/compaction_policy.h [hunk 9]
```diff
@@ -41,4 +41,6 @@ double cumulative_compaction_score(const TabletMetadataPB& metadata);
 
 double base_compaction_score(const TabletMetadataPB& metadata);
 
+double primary_compaction_score(const TabletMetadataPB& metadata);
```
Comment (sduzh) [inline be/src/storage/lake/compaction_policy.h:44]: Perhaps we should only expose an interface for calculating compaction scores:
compaction_policy.h
```
double compaction_score(const TabletMetadataPB& metadata);
``` 

compaction_policy.cpp
```
double compaction_score(const TabletMetadataPB& metadata) {
    if (metadata.keys_type() == PRIMARY_KEYS) {
        return primary_compaction_score(metadata);
    } else {
        return std::max(cumulative_compaction_score(metadata), base_compaction_score(metadata));
    }
}
```

File: be/src/storage/lake/general_tablet_writer.cpp [hunk 10]
```diff
@@ -25,7 +25,10 @@
 
 namespace starrocks::lake {
 
-GeneralTabletWriter::GeneralTabletWriter(Tablet tablet) : _tablet(tablet) {}
+GeneralTabletWriter::GeneralTabletWriter(Tablet tablet, std::shared_ptr<const TabletSchema>& tschema)
+        : _tablet(tablet) {
+    _schema = tschema;
+}
```
Comment (sduzh) [inline be/src/storage/lake/general_tablet_writer.cpp:31]:
```suggestion
GeneralTabletWriter::GeneralTabletWriter(Tablet tablet, std::shared_ptr<const TabletSchema> tschema)
        : _tablet(tablet), _schema(std::move(schema)) {
}
```

File: be/src/storage/lake/delta_writer.cpp [hunk 11]
```diff
@@ -140,21 +141,16 @@ class DeltaWriterImpl {
...
         ASSIGN_OR_RETURN(_tablet_schema, tablet.get_schema());
         RETURN_IF_ERROR(handle_partial_update());
-        if (_tablet_schema->keys_type() == KeysType::PRIMARY_KEYS) {
-            _rowset_txn_meta = std::make_unique<RowsetTxnMetaPB>();
-            ASSIGN_OR_RETURN(_tablet_writer, tablet.new_writer(_rowset_txn_meta.get(), _tablet_schema));
-        } else {
-            ASSIGN_OR_RETURN(_tablet_writer, tablet.new_writer());
-        }
+        ASSIGN_OR_RETURN(_tablet_writer, tablet.new_writer());
+        _tablet_writer->set_tablet_schema(_tablet_schema);
```
Comment (sduzh) [inline be/src/storage/lake/delta_writer.cpp:153]:
```suggestion
        if (_partial_update_tablet_schema != nullptr) {
            _tablet_writer->set_tablet_schema(_partial_update_tablet_schema);
        }
```

File: be/src/storage/lake/tablet_writer.h [hunk 12]
```diff
@@ -72,6 +74,12 @@ class TabletWriter {
 
     // The total number of rows have been written.
     virtual int64_t num_rows() const = 0;
+
+    //  return the rowset txn meta generated by tablet writer
+    virtual RowsetTxnMetaPB* rowset_txn_meta() = 0;
+
+    // allow to set custom tablet schema for writer, used in partial update
+    virtual void set_tablet_schema(std::shared_ptr<const TabletSchema>& tschema) = 0;
```
Comment (sduzh) [inline be/src/storage/lake/tablet_writer.h:82]: It's a little strange to define such a simple setter as virtual function. Maybe we can define `_schema` as a protected field of the base class.

File: be/src/storage/lake/update_compaction_state.cpp [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        pk_cols.resize(rowset->num_segments());
+    }
+}
+
+CompactionState::~CompactionState() {}
+
+Status CompactionState::load_segments(Rowset* rowset, const TabletSchema& tablet_schema, uint32_t segment_id) {
+    if (segment_id >= pk_cols.size() && pk_cols.size() != 0) {
+        std::string msg = strings::Substitute("Error segment id: $0 vs $1", segment_id, pk_cols.size());
+        LOG(WARNING) << msg;
```
Comment (sduzh) [inline be/src/storage/lake/update_compaction_state.cpp:37]: Question: why not just `LOG(WARNING) << "Error segment id: " << segment_id << " vs " << pk_cols.size();`

File: be/src/storage/lake/update_compaction_state.cpp [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+    for (size_t i = 0; i < tablet_schema.num_key_columns(); i++) {
+        pk_columns.push_back(static_cast<uint32_t>(i));
+    }
+
+    VectorizedSchema pkey_schema = ChunkHelper::convert_schema(tablet_schema, pk_columns);
+
+    std::unique_ptr<Column> pk_column;
+    if (!PrimaryKeyEncoder::create_column(pkey_schema, &pk_column).ok()) {
+        CHECK(false) << "create column for primary key encoder failed";
+    }
```
Comment (sduzh) [inline be/src/storage/lake/update_compaction_state.cpp:57]:
```suggestion
    CHECK_OK(PrimaryKeyEncoder::create_column(pkey_schema, &pk_column));
```

File: be/src/storage/lake/update_compaction_state.cpp [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+    OlapReaderStatistics stats;
+    auto res = rowset->get_each_segment_iterator(pkey_schema, &stats);
+    if (!res.ok()) {
+        return res.status();
+    }
+
+    auto& itrs = res.value();
+    CHECK(itrs.size() == rowset->num_segments()) << "itrs.size != num_segments";
```
Comment (sduzh) [inline be/src/storage/lake/update_compaction_state.cpp:66]:
```suggestion
    CHECK_EQ(itrs.size(), rowset->num_segments());
```

File: be/src/storage/lake/lake_primary_index.cpp (hunk 8 above)
Comment (luohaha) [reply be/src/storage/lake/lake_primary_index.cpp]: Done

File: be/src/storage/lake/tablet_writer.h (hunk 12 above)
Comment (luohaha) [reply be/src/storage/lake/tablet_writer.h]: Done

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by sduzh — DISMISSED** [review]

File: be/src/storage/lake/compaction_policy.cpp [hunk 16]
```diff
@@ -59,6 +151,15 @@ double base_compaction_score(const TabletMetadataPB& metadata) {
     return cumulative_point - 1;
 }
 
+double primary_compaction_score(const TabletMetadataPB& metadata) {
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/compaction_policy.cpp:154]: Yes, lake primary table and local primary table are different in many ways.

File: be/src/storage/lake/compaction_policy.cpp [hunk 17]
```diff
@@ -37,6 +40,95 @@ class BaseAndCumulativeCompactionPolicy : public CompactionPolicy {
...
+    explicit PrimaryCompactionPolicy(TabletPtr tablet) : _tablet(std::move(tablet)) {}
+    ~PrimaryCompactionPolicy() override = default;
+
+    StatusOr<std::vector<RowsetPtr>> pick_rowsets(int64_t version) override;
+
+private:
+    static const size_t kCompactionResultBytesThreashold = 1000000000;
+    static const size_t kCompactionResultRowsThreashold = 10000000;
+
+    double calc_compaction_score(const RowsetStat& stats);
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/compaction_policy.cpp]: Done

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by decster — DISMISSED** [review]

**Comment by luohaha** [comment]
run starrocks_fe_unittest

**Review by decster — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by sduzh — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by luohaha** [comment]
run starrocks_admit_test

**Comment by luohaha** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by luohaha** [comment]
run starrocks_admit_test

**Pull request merged**